### ISSUE TITLE
DNS failures for internal AWS URLs

### ISSUE THREAD

**Issue by jkinkead** [kubernetes/dns#85]
I'm seeing intermittent DNS failures, but only for the URLs that AWS assigns my machines (e.g. ` ip-10-16-25-61.us-west-2.compute.internal`).

External URLs (e.g. `google.com`) resolve fine, as do URLs for services (e.g. `kubernetes.default`).

Setup:
Two-replica DNS behind a service IP, configured with `kops`. Each pod is running using [this template](https://github.com/kubernetes/kops/blob/6201e924c7559148cdd10e54a11c6479a9e95a60/upup/models/cloudup/resources/addons/kube-dns.addons.k8s.io/k8s-1.6.yaml.template#L55-L210), which contains `kubedns`, `dnsmasq`, and `dns-sidecar`. All are using version `1.14.1` of the kube-dns images.

The cluster is running Kubernetes version 1.6.0.

Problem description:

The system seems to end up in one of three states:

1. Lookups succeed 100% of the time.

2. The lookup resolves, returning the correct TTL (20). Typically, the first lookup after `kubedns` logs new entries will succeed (each time `kubedns` runs a work cycle, every 5 minutes). Lookups will continue to succeed until the TTL of the entry expires, and will start to fail after that.

3. Lookups fail 100% of the time.

When I've restarted the DNS services, they seem to progress from state 1, through state 2, into state 3, occasionally jumping back to state 1 or 2.

The logs for the containers aren't displaying any errors or other interesting information.

**Comment by jkinkead** [comment]
One more point of context: I'm running pods (instances of the [Datadog agent](https://github.com/DataDog/dd-agent)) which are doing lookups of these URLs regularly, several times a minute.

**Comment by thockin** [comment]
Are those names supposed to be resolvable across VMs?

**Comment by jkinkead** [comment]
Yep, these all resolve correctly (and consistently) through the servers specified in `/etc/resolv.conf`.

Incidentally, `kube-dns` doesn't seem to return the 'authority' section of its responses consistently, either. I'm not sure what that means, or if it's a design choice.

**Comment by jkinkead** [comment]
We're also seeing this now for some other internal DNS entries we've created ourselves - it's not just the AWS-created ones.

**Comment by jkinkead** [comment]
One more datapoint, if it's helpful: I've occasionally seen responses via `dig` that show an 'authority' section but no 'answer' section. This seems especially weird.

**Comment by jkinkead** [comment]
I've convinced myself the problem is how the flags being passed to `dnsmasq` interoperate with private DNS servers, *not* with the core binaries in `kube-dns`.

I'm running a test configuration right now that's showing some promise. I'll close this bug if it continues without error overnight, and follow up with the `kops` people.

Thanks for taking a look!

**Comment by bowei** [comment]
Was there a recent upgrade to 1.6? That is the first version where dnsmasq is configured differently.

**Comment by jkinkead** [comment]
We're running a brand-new cluster, and ran into this issue when we starting running real workloads on it. It's always been 1.6.

**Comment by bowei** [comment]
Are you using the new stub domains feature?

**Comment by jkinkead** [comment]
I don't believe so. We're using `kops` to do setup - see [the template file I linked at the top of the bug](https://github.com/kubernetes/kops/blob/6201e924c7559148cdd10e54a11c6479a9e95a60/upup/models/cloudup/resources/addons/kube-dns.addons.k8s.io/k8s-1.6.yaml.template).

My current theory is that my bug is caused by our `/etc/resolv.conf`, which looks like:

```
nameserver 10.16.0.4
nameserver 8.8.8.8
search internal-name us-west-2.compute.internal
```
`dnsmasq` will use the configured nameservers to do lookups - but the problem is that it will choose the server arbitrarily / randomly, and caches negative lookups. If it picks Google's DNS server (`8.8.8.8`) for resolving internal URLs, it will get a negative result, and return & save that entry.

My current solution is to add the `-o` flag to the `dnsmasq` startup flags, which tells it to use the DNS servers in-order. Our internal nameserver will normally resolve all DNS entries fine, so this is the configuration we want. I *think* this is a much more sensible default than without the flag.

**Comment by bowei** [comment]
It might be worthwhile to take a look at http://blog.kubernetes.io/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes.html.

You can send all requests for `*.compute.internal` to `10.16.0.4` and keep `8.8.8.8` as the upstream nameserver.

**Comment by jkinkead** [comment]
Awesome, thank you - that's extremely helpful.

**Comment by jkinkead** [comment]
I'm closing this in favor of the new `kops` bug.

**Comment by justinsb** [comment]
So this isn't the default AWS configuration, AFAICT, in that this will only happen if you explicitly add 8.8.8.8 to the AWS DHCP options.  But it sounds like that is safe with "normal" DNS, but not safe with dnsmasq, because of the randomization?

Is that right @bowei ?  If so we can indeed add workarounds using stub domains, but this seems like it will hit a lot of people in the real world - and not just with AWS.  And you can have private Route53 domains on AWS, so it isn't a simple finite set like `*.compute.internal`

**Comment by bowei** [comment]
It's never a good idea to mix split horizon name servers in your resolution path. Some libc query sequentially, round robin, musl send queries in parallel (!).

**Comment by NiuZhuang** [comment]
> It might be worthwhile to take a look at https://blog.kubernetes.io/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes.html.
> 
> You can send all requests for `*.compute.internal` to `10.16.0.4` and keep `8.8.8.8` as the upstream nameserver.

This link is 404 now. 
Below is the updated url:
https://kubernetes.io/blog/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes/
